runtime/runtime_events.c

```diff
@@ -717,7 +717,13 @@ CAMLprim value caml_runtime_events_user_register(value event_name,
   Field(event, 3) = event_tag;
 
 
-  caml_plat_lock_non_blocking(&user_events_lock);
+  /* Pre-allocate to avoid STW while holding [user_events_lock]. */
+  list_item = caml_alloc_small(2, 0);
+
+  /* [user_events_lock] can be acquired during STW, so we must use
+     caml_plat_lock_blocking and be careful to avoid triggering any
+     STW while holding it */
+  caml_plat_lock_blocking(&user_events_lock);
```

Comment:
@gadmm I wonder if the following alternative fix would work:
- take the lock `non_blocking` for all the critical sections outside the STW code
- in the STW section, do *not* take `users_events_lock`, with a barrier after the user_events-using section to guarantee that no mutators run in parallel

More generally, I think that we can have resources that use non-blocking locks in mutators and are followed by a barrier in STW.